test/unit/test_task.c

```diff
@@ -0,0 +1,246 @@
+#include <rz_core.h>
+#include "minunit.h"
+
+typedef struct shared_ctx_t {
+	RzStrBuf log;
+	RzThreadLock *free_lock;
+} SharedCtx;
+
+typedef struct local_ctx_t {
+	SharedCtx *shared;
+	int id;
+	int freed;
+	unsigned int want_to_consume;
+	bool breaked;
+} LocalCtx;
+
+static void ctx_switch(RzCoreTask *next, void *user) {
+	SharedCtx *ctx = user;
+	LocalCtx *local_ctx = next->runner_user;
+	if (local_ctx) {
+		rz_strbuf_appendf (&ctx->log, "Context switch to %d\n", local_ctx->id);
+	} else {
+		rz_strbuf_appendf (&ctx->log, "Context switch to main\n");
+	}
+}
+
+static void task_break(RzCoreTask *task, void *user) {
+	LocalCtx *ctx = task->runner_user;
+	ctx->breaked = true;
+}
+
+static void runner(RzCoreTaskScheduler *sched, void *user) {
+	LocalCtx *ctx = user;
+	rz_strbuf_appendf (&ctx->shared->log, "Runner %d start\n", ctx->id);
+	while (ctx->want_to_consume) {
+		if (ctx->breaked) {
+			rz_strbuf_appendf (&ctx->shared->log, "Runner %d was breaked!\n", ctx->id);
+			break;
+		}
+		rz_strbuf_appendf (&ctx->shared->log, "Runner %d consume %u\n", ctx->id, ctx->want_to_consume);
+		ctx->want_to_consume--;
+		rz_core_task_yield (sched);
+	}
+	rz_strbuf_appendf (&ctx->shared->log, "Runner %d done\n", ctx->id);
+}
+
+static void local_ctx_free(void *user) {
+	LocalCtx *ctx = user;
+	SharedCtx *shared_ctx = ctx->shared;
+	rz_th_lock_enter (shared_ctx->free_lock);
+	ctx->freed++;
+	rz_th_lock_leave (shared_ctx->free_lock);
+}
+
+// Busy-wait for task threads to be started and in the queue
+static void wait_for_tasks_enqueued(RzCoreTaskScheduler *sched, size_t count) {
+	while (true) {
+		rz_th_lock_enter (sched->lock);
+		size_t queued = rz_list_length (sched->tasks_queue);
+		rz_th_lock_leave (sched->lock);
+		if (queued == count) {
+			break;
+		}
+		rz_sys_usleep (1000);
+	}
+}
+
+static bool test_task(void) {
+	SharedCtx shared_ctx = { 0 };
+	rz_strbuf_init (&shared_ctx.log);
+	shared_ctx.free_lock = rz_th_lock_new (false);
+
+	RzCoreTaskScheduler sched;
+	rz_core_task_scheduler_init (&sched, ctx_switch, &shared_ctx, task_break, &shared_ctx);
+	rz_core_task_sync_begin (&sched);
+
+	LocalCtx a_ctx = {
+		.shared = &shared_ctx,
+		.id = 1,
+		.want_to_consume = 3
+	};
+	RzCoreTask *a = rz_core_task_new (&sched, runner, local_ctx_free, &a_ctx);
+	int a_id = a->id;
+	rz_core_task_enqueue (&sched, a);
+	a = NULL; // ownership moved to the scheduler, don't touch anymore!
```

Comment:
We will see this pattern often probably. Might be useful to create a macros or a function for this.

Reply:
What pattern do you mean exactly?